lib/read.js

```diff
@@ -182,68 +174,81 @@ var Read = Juttle.proc.source.extend({
             buckets.push(intermediate_bucket);
 
             _.each(buckets, function(b, i) {
-                logger.debug('time bucket ' + i, new Date(b.valueOf()).toISOString());
+                self.logger.debug('time bucket ' + i, new Date(b.valueOf()).toISOString());
             });
             return buckets;
         }
         this.batchBuckets = get_buckets();
-    },
+    }
 
-    getNextLimit: function () {
-        return this.maxSize === undefined ? this.fetchSize : Math.min(this.fetchSize, this.maxSize - this.total_emitted_points);
-    },
+    getNextLimit() {
+        return Math.min(this.fetchSize, this.queueSize, this.maxSize - this.total_emitted_points);
+    }
 
     //Query execution
 
-    start: function() {
-        var self = this;
+    read(from, to, limit, state) {
+        var queryExecutionPromise;
 
-        if (this.debugOption) {
-            this.handleDebug();
-            return;
-        }
+        this.queueSize = limit;
 
-        function executeQuery() {
-            if (self.raw) {
-                return self.baseQuery
-                .then(function(res) {
-                    return self.handleRawResponse(res);
-                })
-                .then(function(points) {
-                    return self.formatAndSend(points);
-                });
+        if (this.raw) {
+            queryExecutionPromise = this.baseQuery
+            .then((res) => {
+                return this.handleRawResponse(res);
+            })
+            .then((points) => {
+                return this.formatAndSend(points);
```

Comment:
I might rename this as you're not really sending any longer. But I don't feel that strongly about it.